chore: move metadata to pyproject.toml

SauravMaheshkar opened the pull request:
`pyproject.toml` is the new standard for packaging python packages (first introduced in [PEP 518](https://peps.python.org/pep-0518/) and later expanded in [PEP 517](https://peps.python.org/pep-0517/), [PEP 621](https://peps.python.org/pep-0621/) and [PEP 660](https://peps.python.org/pep-0660/)). This PR proposes to switch the metadata declaration from the legacy `setup.cfg` in favour of `pyproject.toml`. This would also allow for a better minimal project structure.

SauravMaheshkar commented:
@nkoep Request for Review

nkoep commented on pyproject.toml:
```diff
@@ -2,9 +2,97 @@
 
+[project]
+name = "pymanopt"
+description = "Toolbox for optimization on Riemannian manifolds with support for automatic differentiation"
```
```suggestion
description = """\
  Toolbox for optimization on Riemannian manifolds with support for automatic \
  differentiation\
"""
```

SauravMaheshkar replied:
Addressed in edda1b20cecb5a0b2ee767471ba284f301713545

nkoep commented on pyproject.toml:
```diff
@@ -2,9 +2,97 @@
+  {name = "Niklas Koep"},
+  {name = "Sebastian Weichwald"},
+]
+readme="README.md"
```
```suggestion
readme = "README.md"
```

SauravMaheshkar replied:
Addressed in edda1b20cecb5a0b2ee767471ba284f301713545

nkoep commented on pyproject.toml:
```diff
@@ -2,9 +2,97 @@
+]
+readme="README.md"
+license = { text = "BSD 3-Clause License"}
+requires-python=">=3.7"
+keywords=[
```
```suggestion
requires-python = ">=3.7"
keywords = [
```

SauravMaheshkar replied:
Addressed in edda1b20cecb5a0b2ee767471ba284f301713545

nkoep reviewed (changes requested):
Hey, thanks a lot for the PR! I've been meaning to get rid of the `setup.cfg` file as well. I just added a bunch of pedantic style comments. If you don't mind, could you also change all indents to 2 spaces for consistency? :bow:

SauravMaheshkar commented on pyproject.toml:
```diff
@@ -2,9 +2,97 @@
+  {name = "Sebastian Weichwald"},
+]
+readme="README.md"
+license = { text = "BSD 3-Clause License"}
```
Addressed in edda1b20cecb5a0b2ee767471ba284f301713545

SauravMaheshkar commented on pyproject.toml:
```diff
@@ -2,9 +2,97 @@
+    "scipy>=1.0,<1.10",
+]
+[project.urls]
+homepage="https://pymanopt.org"
```
Addressed in edda1b20cecb5a0b2ee767471ba284f301713545

SauravMaheshkar commented:
> Hey, thanks a lot for the PR! I've been meaning to get rid of the `setup.cfg` file as well. I just added a bunch of pedantic style comments. If you don't mind, could you also change all indents to 2 spaces for consistency? 🙇

Hello @nkoep thanks a lot for the feedback. I addressed the requested changes in edda1b20cecb5a0b2ee767471ba284f301713545. Looking forward to contributing more 😄

nkoep reviewed (approved):
Thanks for the swift updates! LGTM!

SauravMaheshkar commented:
@nkoep Do you think we should try and add a `pip install -U pip setuptools wheel` in the CI ? This repository does use the older version of `setup-python` in CI. I can make a quick PR to update it.

nkoep commented:
> @nkoep Do you think we should try and add a `pip install -U pip setuptools wheel` in the CI ? This repository does use the older version of `setup-python` in CI. I can make a quick PR to update it. Or maybe update this PR

I think this should be enough:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index bcb5f7e..5f885e4 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,5 +1,5 @@
 [build-system]
-requires = ["pip==21.3.1", "setuptools==59.6.0", "setuptools-scm[toml]>=6.2"]
+requires = ["pip==22.3.1", "setuptools==65.6.3", "setuptools-scm[toml]>=6.2"]
 build-backend = "setuptools.build_meta"
 ```
Feel free to add it to this PR.

SauravMaheshkar commented:
Can I add 

```lang-toml
version = "2.1.1"
```

nkoep commented on pyproject.toml:
```diff
@@ -1,10 +1,101 @@
 build-backend = "setuptools.build_meta"
 
+[project]
+name = "pymanopt"
```
This seems necessary to get `setuptools-scm` to work.
```suggestion
name = "pymanopt"
dynamic = ["version"]
```
